### ISSUE TITLE
Suggest `x <$ y` instead of `const x <$> y`

### ISSUE THREAD

**Issue by foxbenjaminfox** [ndmitchell/hlint#546]
More than once I've found myself writing something along the lines of `const x <$> y`, and only later realizing that I can replace that with `x <$ y`. This is definitely something that hlint could suggest.

The hint for this would be
```yaml
- hint: {lhs: const x <$> y, rhs: x Data.Functor.<$ y}
```
There probably should also be the equivalent hints for `pure x <$> y` and `return x <$> y`.

Perhaps for completeness's sake we'd also want to suggest `x $> y` instead of `x <&> const y`, and likewise for `x <&> pure y` and `x <&> return y`.

If you're OK with these, I'll open a PR. I'm not so sure if the flipped versions comes up often enough to be worthwhile, so if you don't think they're worth it I can leave them out.

**Comment by ndmitchell** [comment]
`<$` is in the Prelude, so no need to qualify even. Is the `pure`/`return` equivalent to const in these cases, being a `->` monad? Are pure/return there common, if they are just const? Can the `pure`/`return` there be in a different monad, such that the simplification is no longer valid? (My gut feel is yes, but maybe the application after stops that being the case?)

I'm happy with these, even the flipped versions. Query on if the pure/return playing the role of const, but only on that aspect.

**Comment by foxbenjaminfox** [comment]
About `<$` being in the Prelude: I was basing the qualification based on [your comment](https://github.com/ndmitchell/hlint/issues/395#issuecomment-338491540) on #395, so I'm confused, because `<$` was already in the Prelude at that point (since GHC 7.10 in 2015, according to [this Trac issue](https://ghc.haskell.org/trac/ghc/ticket/10113)).

About `pure`/`return`: Yes, they're playing the role of `const` here, and only `const`. The first argument of `<$>` (or of `fmap`) is a function, so if the first argument to `<$>` is `pure x`, then `pure x` has to be a function. `pure x` has the type `Application f => f a`, and the only way that type can be a function is if `f` is specialized to `(->) r` (for some `r`).

**Comment by foxbenjaminfox** [comment]
I just realized that I was confused between `<$` and `$>`: it turns out that while `<$` is in the Prelude, for some reason `$>` isn't. So that's why `$>` needed qualification in #395, whereas `<$` doesn't. So never mind about that then.

**Comment by ndmitchell** [comment]
Do people really use pure/return instead of const? I know they could, but does it happen in practice?

**Comment by foxbenjaminfox** [comment]
Probably not so much in the general case.

In this specific context it may come up more than it might usually, because both `pure x <* y` and `pure x <$> y` mean the same thing, `x <$ y`. This is no more than a coincidence: in the first case `pure` is `const`, and in the second it is the `pure` of whatever Applicative instance `y` has, but both can be simplified to the same expression, `x <$ y`. (There already is a hint for the second one, and my PR adds a hint for the first one.)

This is the main reason why having hints for `pure` and `return` here seems useful. Someone who writes `pure x <$> y` by mistake, meaning to write `pure x <* y`, won't have their mistake pointed out to them, because both expressions are equivalent, although `pure x <$> y` is probably a worse way to write it, because it's not at all obvious how that works (with `pure` being specialized here to `const`.)

HLint could have a hint to suggest `pure x <* y` instead of `pure x <$> y`, as that kind of thing is exactly the kind of mistake HLint is supposed to hint about. But that itself isn't a good hint, because when the user fixes their `pure x <$> y` to `pure x <* y` they'd immediately get another hint from HLint telling them to change `pure x <* y` to `x <$ y`.

So better to have the hint for `pure x <$> y` suggest `x <$ y` directly. In any case I think we'd want a hint for `pure x <$> y`, because it's confusing and unintuitive about what it's doing, even if it technically works.

**Comment by ndmitchell** [comment]
Fair enough, since you implemented it, I'm happy enough to give you the discretion to do what suits you :)
